src/MultilineStrings.jl

```diff
@@ -0,0 +1,253 @@
+
+    # Note: Using '\0` to indicate undefined
+    yaml_chomp = false
+    style_char, chomp_char = if indicators_len == 2
+        indicators
+    elseif indicators_len == 1
+        ind = indicators[1]
+        if ind in ('f', 'l')
+            ind, '\0'
+        elseif ind in ('>', '|')
+            yaml_chomp = true
+            ind, '\0'
+        else
+            '\0', ind
+        end
+    else
+        '\0', '\0'
+    end
+
+    if style_char != '\0' && chomp_char != '\0' && isletter(style_char) ⊻ isletter(chomp_char)
+        throw(ArgumentError("Can't mix YAML style block indicators with letter indicators"))
+    end
+
+    style = if style_char == 'f' || style_char == '>'
+        :folded
+    elseif style_char == 'l' || style_char == '|'
+        :literal
+    elseif style_char == '\0'
+        DEFAULT_STYLE
+    else
+        throw(ArgumentError("Unknown style indicator: $(repr(style_char))"))
+    end
+
+    chomp = if chomp_char == 'c' || yaml_chomp
+        :clip
+    elseif chomp_char == 's' || chomp_char == '-'
+        :strip
+    elseif chomp_char == 'k' || chomp_char == '+'
+        :keep
+    elseif chomp_char == '\0'
+        DEFAULT_CHOMP
+    else
+        throw(ArgumentError("Unknown chomping indicator: $(repr(chomp_char))"))
+    end
+
+    return multiline(str, style, chomp)
+end
+
+function multiline(str::AbstractString, style::Symbol, chomp::Symbol)
+    # Append an additional, non-space, character to force one more iteration of the style
+    # loop
+    str *= ETX
+
+    out = IOBuffer()
+    num_newlines = 0  # The number of newlines at the end of the string
+    prev = curr = '\0'
+
+    # Replace newlines with spaces (folded)
+    if style === :folded
+        # The code below is equivalent to these two regexes:
+        # ```
+        # str = replace(str, r"(?<=\S)\n(?=\S)" => " ")
+        # str = replace(str, r"(?<=\n)\n(?=\S)" => "")
+        # ```
+
+        for next in str
+            if curr == '\n'
+                if !isspace(next) && next != ETX
+                    if prev == '\n'
+                        # Skip last newline in a sequence of sequential blank lines
+                    elseif !isspace(prev)
+                        # Replace a single newline with a space
+                        write(out, ' ')
+                    else
+                        num_newlines += 1
+                    end
+                else
+                    num_newlines += 1
+                end
+            elseif curr != '\0'
+                # Insert newlines which were determined to not be at the end of the string
+                if num_newlines > 0
+                    write(out, "\n" ^ num_newlines)
+                    num_newlines = 0
+                end
+
+                write(out, curr)
+            end
+
+            prev = curr
+            curr = next
+        end
+
+    # Keep newlines (literal)
+    elseif style === :literal
```

Comment:
Should we break these two branches out into seperate methods?
They are really long.

Reply:
I attempted to separate this but doing so also leads me to refactoring the common logic out into a separate function which overall I think makes thing less readable. If I just break these out as is as functions we end up having to copy a bunch of setup shared between these loops.

I think leaving this as is may be the clearest even though it is long